From: Lin Ming <ming.m.lin@intel.com>
Subject: ACPI: Fix mem corruption
References: 350017
Patch-mainline: hopefully soon

Fix a memory overflow bug when copying NULL internal package element object to
external.

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 drivers/acpi/utilities/utobject.c |    2 +-
 drivers/acpi/utils.c              |    6 ++++++
 2 files changed, 7 insertions(+), 1 deletion(-)

Index: linux-2.6.24/drivers/acpi/utilities/utobject.c
===================================================================
--- linux-2.6.24.orig/drivers/acpi/utilities/utobject.c
+++ linux-2.6.24/drivers/acpi/utilities/utobject.c
@@ -432,7 +432,7 @@ acpi_ut_get_simple_object_size(union acp
 	 * element -- which is legal)
 	 */
 	if (!internal_object) {
-		*obj_length = 0;
+		*obj_length = sizeof(union acpi_object);
 		return_ACPI_STATUS(AE_OK);
 	}
 
Index: linux-2.6.24/drivers/acpi/utils.c
===================================================================
--- linux-2.6.24.orig/drivers/acpi/utils.c
+++ linux-2.6.24/drivers/acpi/utils.c
@@ -407,6 +407,12 @@ acpi_evaluate_reference(acpi_handle hand
 			break;
 		}
 
+		if (!element->reference.handle) {
+			printk(KERN_WARNING PREFIX "Invalid reference in"
+			       " package %s\n", pathname);
+			status = AE_NULL_ENTRY;
+			break;
+		}
 		/* Get the  acpi_handle. */
 
 		list->handles[i] = element->reference.handle;
